### ISSUE TITLE
Enable strictNullChecks for datetime package

### ISSUE THREAD

**Issue by giladgray** [palantir/blueprint#325]
I think enabling `strictNullChecks` compiler option for `datetime` package will help us detect and prevent a whole host of potential errors, such as #322. it'll also address #274 by making the typings more accurate (as explicit `null` Dates actually means something in the component).

**Comment by adidahiya** [comment]
I think the compiler option should be enabled for all packages. Updating issue title to reflect this.

**Comment by giladgray** [comment]
Fair enough! Converting `core` will take some time due to its sheer size, hence why I started with `datetime`. Any idea if this will affects typings in a breaking way?

**Comment by adidahiya** [comment]
No, any changes in the generated `.d.ts` files ought to be tractable & compatible with consumers' `tsc` versions. If anything, it will expose bugs in consuming applications (not handling null / undefined values).

**Comment by giladgray** [comment]
> in my ideal world, we also enable the `no-null-keyword` lint rule and any explicit usage of `null` would have to be whitelisted

https://github.com/palantir/blueprint/pull/419#discussion_r94995122

**Comment by giladgray** [comment]
started working on this and quickly ran into some heavy sadness with `setState`: 
- https://github.com/Microsoft/TypeScript/issues/6613
- https://github.com/Microsoft/TypeScript/issues/12793

basically, `this.state` fields should all be required (otherwise you have to cast/check before _using_ any as they're all `X | undefined`), but then we can't `setState()` with partial objects. the solution is `Pick` and @ericanderson has submitted a PR to DT to update react typings accordingly: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/13155.

**Comment by adidahiya** [comment]
Cool, it looks like that PR will be merged in DefinitelyTyped soon enough and I don't mind blocking on it.

**Comment by adidahiya** [comment]
We're good to go on the `setState` enhancements in the React typings now. Pull the latest `@types/react`.

**Comment by brieb** [comment]
If we forbid using `null`, what's the best way to indicate an empty value for controlled input components. Using `undefined` doesn't work because that indicates that the component should be in uncontrolled mode. Some inputs have a logical empty value - like for text inputs, the empty value should be represented as `""`. But, what about for something less clear - like a date input. Should we export an explicit empty value to use in that case? Looks like we do that here https://github.com/palantir/blueprint/blob/master/packages/core/src/components/forms/numericInput.tsx#L138

**Comment by adidahiya** [comment]
(updated your link to be a permalink)

> Should we export an explicit empty value to use in that case?

yeah, this sounds reasonable to me. I would be curious to see how far we can get with avoiding `null` in our public API.

**Comment by ggdouglas** [comment]
The original goal, enabling strictNullChecks for published library packages, is complete. core (2020), table (2022), and datetime (2025) were migrated explicitly; all other library packages inherit strict: true from the shared base tsconfig. The only remaining strictNullChecks: false overrides are in internal apps (docs-app, demo-app, landing-app, table-dev-app), which don’t affect published typings. Remaining cleanup can be tracked separately.
